src/undo-manager.js

```diff
@@ -1,43 +1,73 @@
 define(function () {
+	'use strict';
 
-  'use strict';
+	function UndoManager(limit, undoScopeHost) {
+		var stack = [];
+		var fireEvent = typeof CustomEvent != 'undefined' && undoScopeHost && undoScopeHost.dispatchEvent;
 
-  return function (scribe) {
+		this.position = 0;
+		this.length = 0;
 
-    function UndoManager() {
-      this.position = -1;
-      this.stack = [];
-      this.debug = scribe.isDebugModeEnabled();
-    }
+		this.transact = function (transaction, merge) {
+			if (arguments.length < 2)
+				throw new TypeError('Not enough arguments to UndoManager.transact.');
 
-    UndoManager.prototype.maxStackSize = 100;
+			transaction.execute();
 
-    UndoManager.prototype.push = function (item) {
-      if (this.debug) {
-        console.log('UndoManager.push: %s', item);
-      }
-      this.stack.length = ++this.position;
-      this.stack.push(item);
+			stack.splice(0, this.position);
+			if (merge && this.length)
+				stack[0].push(transaction);
+			else
+				stack.unshift([transaction]);
+			this.position = 0;
 
-      while (this.stack.length > this.maxStackSize) {
-        this.stack.shift();
-        --this.position;
-      }
-    };
+			if (limit && stack.length > limit)
+				this.length = stack.length = limit;
+			else
+				this.length = stack.length;
 
-    UndoManager.prototype.undo = function () {
-      if (this.position > 0) {
-        return this.stack[--this.position];
-      }
-    };
+			if (fireEvent)
+				undoScopeHost.dispatchEvent(new CustomEvent('DOMTransaction', {detail: {transactions: stack[0].slice()}, bubbles: true, cancelable: false}));
+		};
 
-    UndoManager.prototype.redo = function () {
-      if (this.position < (this.stack.length - 1)) {
-        return this.stack[++this.position];
-      }
-    };
+		this.undo = function () {
+			if (this.position < this.length) {
+				for (var i = stack[this.position].length - 1; i >= 0; i--)
+					stack[this.position][i].undo();
+				this.position++;
```

Comment:
When we were discussing this internally we thought it was better to use a reversed stack (append at head) rather than a decrementing index loop.